src/content/content_encoding.rs

```diff
@@ -0,0 +1,147 @@
+//! Specify the compression algorithm.
+
+use crate::content::{Encoding, EncodingProposal};
+use crate::headers::{HeaderName, HeaderValue, Headers, ToHeaderValues, CONTENT_ENCODING};
+
+use std::fmt::{self, Debug};
+use std::ops::{Deref, DerefMut};
+use std::option;
+
+/// Specify the compression algorithm.
+///
+/// # Specifications
+///
+/// - [RFC 7231, section 3.1.2.2: Content-Encoding](https://tools.ietf.org/html/rfc7231#section-3.1.2.2)
+///
+/// # Examples
+///
+/// ```
+/// # fn main() -> http_types::Result<()> {
+/// #
+/// use http_types::Response;
+/// use http_types::content::{ContentEncoding, Encoding};
+/// let mut encoding = ContentEncoding::new(Encoding::Gzip);
+///
+/// let mut res = Response::new(200);
+/// encoding.apply(&mut res);
+///
+/// let encoding = ContentEncoding::from_headers(res)?.unwrap();
+/// assert_eq!(encoding, &Encoding::Gzip);
+/// #
+/// # Ok(()) }
+/// ```
+pub struct ContentEncoding {
+    inner: Encoding,
+}
+
+impl ContentEncoding {
+    /// Create a new instance of `CacheControl`.
+    pub fn new(encoding: Encoding) -> Self {
+        Self { inner: encoding }
+    }
+
+    /// Create a new instance from headers.
+    pub fn from_headers(headers: impl AsRef<Headers>) -> crate::Result<Option<Self>> {
+        let headers = match headers.as_ref().get(CONTENT_ENCODING) {
+            Some(headers) => headers,
+            None => return Ok(None),
+        };
+
+        let mut inner = None;
+
+        for value in headers {
+            for part in value.as_str().trim().split(',') {
+                // Try and parse a directive from a str. If the directive is
+                // unkown we skip it.
+                if let Some(entry) = Encoding::from_str(part) {
+                    inner = Some(entry);
+                }
+            }
+        }
+
+        let inner = inner.expect("Headers instance with no entries found");
```

Comment:
This probably shouldn't panic?

Reply:
This should indeed never panic, but still adding the message just in case we ever mess up that constraint. Wish there was a better way to express: "If we have a list, there is always at least one item. We never create an empty list, that's `None` instead".

But ah well; can't have it all.